xiplot/tabs/plugins.py

```diff
@@ -0,0 +1,406 @@
+        )
+
+    @staticmethod
+    def create_layout_globals():
+        return html.Div(
+            [
+                html.Div(
+                    id="plugin-tab-notify-container", style={"display": "none"}
+                ),
+                html.Div(
+                    id="plugin-tab-upload-notify-container",
+                    style={"display": "none"},
+                ),
+                dcc.Input(
+                    id="plugin-tab-load-progress",
+                    style={"display": "none"},
+                    type="text",
+                ),
+                dcc.Input(
+                    id="plugin-tab-load-success",
+                    style={"display": "none"},
+                    type="text",
+                ),
+                dcc.Input(
+                    id="plugin-tab-load-exception",
+                    style={"display": "none"},
+                    type="text",
+                ),
+            ]
+        )
+
+
+def get_loaded_plugin_options():
+    plugins = defaultdict(set)
+
+    for (
+        kind,
+        name,
+        path,
+        plugin,
+    ) in get_all_loaded_plugins():
+        plugins[path.split(":")[0].split(".")[0]].add(kind)
+
+    plugin_options = []
+
+    for plugin, kinds in plugins.items():
+        if len(kinds) > 0:
+            kinds = f": {', '.join(kinds)}"
+        else:
+            kinds = ""
+
+        plugin_options.append(
+            {
+                "label": f"{plugin}{kinds}",
+                "value": plugin,
+                "disabled": True,
+            }
+        )
+
+    return plugin_options
+
+
+def get_plugin_filepaths(dir_path=""):
+    try:
+        return sorted(
+            (
+                fp
+                for fp in Path(dir_path).iterdir()
+                if fp.is_file() and fp.suffix == ".whl"
+            ),
+            reverse=True,
+        )
+
+    except FileNotFoundError:
+        return []
+
+
+def install_local_plugin(plugin_path: str):
+    import asyncio
+
+    try:
+        import micropip
+    except ImportError:
+        raise NotImplementedError(
+            "Loading new plugins is only supported in WASM"
+        )
+
+    plugin_path = Path(plugin_path).resolve()
+
+    asyncio.ensure_future(
+        micropip.install(f"emfs:{str(plugin_path)}")
+    ).add_done_callback(
+        partial(__micropip_install_callback, try_remove_file=plugin_path)
+    )
+
+
+def install_remote_plugin(plugin_source: str):
+    import asyncio
+
+    try:
+        import micropip
+    except ImportError:
+        raise NotImplementedError(
+            "Loading new plugins is only supported in WASM"
+        )
+
+    if len(plugin_source.split()) != 1:
+        raise ValueError("Plugin source must be a URL or PyPi package name")
+
+    if (
+        Path(plugin_source).exists()
+        and Path(plugin_source).name != plugin_source
+    ):
+        raise ValueError("Plugin source must be a URL or PyPi package name")
```

Comment:
Is there a reason for not letting micropip handle the parsing of the string (and throw a more detailed error)?